File: 99.NumberTheoreticTransformAlgo.py
# The Number Theoretic Transform (NTT) is the modular 
# arithmetic version of the Fast Fourier Transform (FFT).
# It’s used to perform polynomial multiplication efficiently under modular 
# arithmetic, avoiding floating-point errors.

# Detailed Explanation

# The NTT works similarly to FFT but uses a primitive root of 
# unity modulo a prime number instead of complex roots of unity.
# It allows us to:

# Multiply large polynomials efficiently (O(n log n) instead of O(n²))

# Work entirely with integers under a prime modulus p = k * 2^n + 1

# For example, modulus p = 998244353 and primitive root 
# g = 3 are often used in competitive programming.

MOD = 998244353
ROOT = 3

def power(a, b, mod):
    res = 1
    while b > 0:
        if b & 1:
            res = res * a % mod
        a = a * a % mod
        b >>= 1
    return res

def ntt(a, invert):
    n = len(a)
    j = 0
    for i in range(1, n):
        bit = n >> 1
        while j & bit:
            j ^= bit
            bit >>= 1
        j ^= bit
        if i < j:
            a[i], a[j] = a[j], a[i]

    len_ = 2
    while len_ <= n:
        wlen = power(ROOT, (MOD - 1) // len_, MOD)
        if invert:
            wlen = power(wlen, MOD - 2, MOD)
        for i in range(0, n, len_):
            w = 1
            for j in range(i, i + len_ // 2):
                u = a[j]
                v = a[j + len_ // 2] * w % MOD
                a[j] = (u + v) % MOD
                a[j + len_ // 2] = (u - v + MOD) % MOD
                w = w * wlen % MOD
        len_ <<= 1

    if invert:
        inv_n = power(n, MOD - 2, MOD)
        for i in range(n):
            a[i] = a[i] * inv_n % MOD

def multiply(poly1, poly2):
    n = 1
    while n < len(poly1) + len(poly2):
        n <<= 1
    f = poly1 + [0] * (n - len(poly1))
    g = poly2 + [0] * (n - len(poly2))

    ntt(f, False)
    ntt(g, False)
    for i in range(n):
        f[i] = f[i] * g[i] % MOD
    ntt(f, True)
    return f

A = [1, 2, 3]
B = [4, 5, 6]

res = multiply(A, B)
print("Resultant Polynomial Coefficients:", res)

File: 99.NumberTheoreticTransformAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio       */
/* -----------------------------------------------------------------------  */

import java.io.*;

public class Main {

    static final int MOD = 998244353;   // Prime modulus (k*2^n + 1)
    static final int ROOT = 3;          // Primitive root of MOD

    // -------------------- Fast Exponentiation -------------------- //
    static long power(long a, long b, int mod) {
        long res = 1;
        while (b > 0) {
            if ((b & 1) == 1) res = res * a % mod;  // If exponent is odd
            a = a * a % mod;                        // Square base
            b >>= 1;                                // Divide exponent by 2
        }
        return res;
    }

    // -------------------- Number Theoretic Transform -------------------- //
    static void ntt(long[] a, boolean invert) {
        int n = a.length;

        // Bit-reversal permutation
        for (int i = 1, j = 0; i < n; i++) {
            int bit = n >> 1;
            for (; (j & bit) != 0; bit >>= 1)
                j ^= bit;
            j ^= bit;
            if (i < j) {
                long temp = a[i];
                a[i] = a[j];
                a[j] = temp;
            }
        }

        // Cooley–Tukey iterative NTT
        for (int len = 2; len <= n; len <<= 1) {
            long wlen = power(ROOT, (MOD - 1) / len, MOD);
            if (invert) wlen = power(wlen, MOD - 2, MOD); // Inverse root

            for (int i = 0; i < n; i += len) {
                long w = 1;
                for (int j = 0; j < len / 2; j++) {
                    long u = a[i + j];
                    long v = a[i + j + len / 2] * w % MOD;
                    a[i + j] = (u + v) % MOD;
                    a[i + j + len / 2] = (u - v + MOD) % MOD;
                    w = w * wlen % MOD;
                }
            }
        }

        // Scaling for inverse NTT
        if (invert) {
            long invN = power(n, MOD - 2, MOD);
            for (int i = 0; i < n; i++) a[i] = a[i] * invN % MOD;
        }
    }

    // -------------------- Polynomial Multiplication -------------------- //
    static long[] multiply(long[] a, long[] b) {
        int n = 1;
        while (n < a.length + b.length) n <<= 1;

        long[] f = new long[n];
        long[] g = new long[n];

        // Copy input coefficients
        for (int i = 0; i < a.length; i++) f[i] = a[i];
        for (int i = 0; i < b.length; i++) g[i] = b[i];

        ntt(f, false);  // Forward NTT
        ntt(g, false);  // Forward NTT

        for (int i = 0; i < n; i++) f[i] = f[i] * g[i] % MOD;  // Point-wise multiplication

        ntt(f, true);   // Inverse NTT to get final coefficients
        return f;
    }

    // -------------------- Main Function -------------------- //
    public static void main(String[] args) throws IOException {
        BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out));

        long[] A = {1, 2, 3};   // Polynomial 1: 1 + 2x + 3x²
        long[] B = {4, 5, 6};   // Polynomial 2: 4 + 5x + 6x²

        long[] res = multiply(A, B);

        out.write("Resultant Polynomial Coefficients: ");
        for (long v : res) out.write(v + " ");
        out.newLine();

        out.flush();
    }
}
